### PR TITLE
Support configuration of aggregation function arguments

### PR DESCRIPTION
**Pull request by atumbde**

### CODE REVIEW

**Comment by antumbde** [comment]
@kishoreg @mayankshriv @Jackie-Jiang : Take a look at this. We configure specific aggregation functions in type-safe manner without creating large context objects or passing high level objects like segmentMEtadata to lower layers.

This is WIP only to validate concept. I will clean up the code and move inline anonymous visitor to proper place.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunction.java [hunk 1]
```diff
@@ -37,6 +37,8 @@
     PERCENTILEEST_QUANTILEDIGEST
   }
 
+  void accept(AggregationFunctionVisitorBase visitor);
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunction.java]: If we are using visitor pattern, should we define the visitable interface?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java [hunk 2]
```diff
@@ -49,13 +52,12 @@
    * @param projectionOperator Projection operator.
    * @param numTotalRawDocs Number of total raw documents.
    */
-  public AggregationOperator(List<AggregationInfo> aggregationsInfoList, MProjectionOperator projectionOperator,
+  public AggregationOperator(AggregationFunctionContext[] aggrFuncContextArray, MProjectionOperator projectionOperator,
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: Until we find a better way, should numTotalRawDocs go inside the context?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunction.java (hunk 1 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunction.java]: The AggregationFunctionVisitorBase is a no-op base class that allows creation of specialized classes overriding only relevant methods. Interface will require every derived class to implement all the methods.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java (hunk 2 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: I can move it

**Comment by antumbde** [comment]
Submitted complete set of changes for this PR

**Comment by antumbde** [comment]
For history/context, see PR #865

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java [hunk 3]
```diff
@@ -46,23 +45,23 @@
...
    *
    * @param aggregationsInfoList List of AggregationInfo (contains context for applying aggregation functions).
+   * @param executor Aggregation Executor
    * @param projectionOperator Projection operator.
    * @param numTotalRawDocs Number of total raw documents.
    */
-  public AggregationOperator(List<AggregationInfo> aggregationsInfoList, MProjectionOperator projectionOperator,
-      long numTotalRawDocs) {
+  public AggregationOperator(List<AggregationInfo> aggregationsInfoList, AggregationExecutor executor,
+      MProjectionOperator projectionOperator,long numTotalRawDocs) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: Missing what space after 'projectionOperator,' -> Can you format the file as per the styling?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java [hunk 4]
```diff
@@ -46,23 +45,23 @@
    * Constructor for the class.
    *
    * @param aggregationsInfoList List of AggregationInfo (contains context for applying aggregation functions).
+   * @param executor Aggregation Executor
    * @param projectionOperator Projection operator.
    * @param numTotalRawDocs Number of total raw documents.
    */
-  public AggregationOperator(List<AggregationInfo> aggregationsInfoList, MProjectionOperator projectionOperator,
-      long numTotalRawDocs) {
+  public AggregationOperator(List<AggregationInfo> aggregationsInfoList, AggregationExecutor executor,
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: So planning phase will allocate executor now? If so, what's the motivation for this?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java (hunk 3 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: done

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java (hunk 4 above)
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: This moves operator wiring logic to the planner. Given that we have interface-impl pattern for AggregationExecutor, decision to instantiate executors will be based on segment metadata or brokerRequest. That decision should be made by planner. Operators then have single responsibility to to perform specific operation using the inputs.

**Pull request merged**